### PR TITLE
Added Support for Custom Turret Controllers and Targeting Options

### LINKED ISSUE #204: [Feature Request] Turret Target Option Support
**Is your feature request related to a problem? Please describe.**
They recently added the ability to select system types to target.  We should update the Target property of turrets to accept a String, and have it get/set the type (default, weapons, power, propulsion).

As part of this, it might be time to add a global custom property word to value mapping.  The case of "I want to use words to set a property value to a specific number" is getting more common.  Other examples include FlightMode, Text alignment, etc.

**Describe the solution you'd like**
```
tell "My Turrets" to target "weapons"
```

### PR DESCRIPTION
**Pull request by MerlinofMines**
This PR adds support for Custom Turret Controllers, along with some amended and additional properties for turrets including azimuth/elevation control and basic targeting options support.

This PR resolves #204 and #212

### CODE REVIEW

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                AddDirectionHandlers(Property.VELOCITY, Direction.RIGHT,
+                    TypeHandler(NumericHandler(b => b.VelocityMultiplierAzimuthRpm, (b,v) => b.VelocityMultiplierAzimuthRpm = v), Direction.LEFT, Direction.RIGHT),
+                    TypeHandler(NumericHandler(b => b.VelocityMultiplierElevationRpm, (b,v) => b.VelocityMultiplierElevationRpm = v), Direction.UP, Direction.DOWN));
+                defaultPropertiesByPrimitive[Return.NUMERIC] = Property.RANGE;
+            }
+
+            IEnumerable<IMyUserControllableGun> GetGuns(IMyTurretControlBlock block) {
+                var tools = NewList<IMyFunctionalBlock>();
+                block.GetTools(tools);
+                return tools.Where(t => t is IMyUserControllableGun).Select(t => (IMyUserControllableGun)t);
```
Comment (jgersti) [inline EasyCommands/BlockHandlers/TurretBlockHandlers.cs:49]: The Pattern `.Where(t => t is T).Select(t -> (T)t)` can always be replaced by `OfType<T>()`

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                AddPropertyHandler(Property.LOCKED, TerminalPropertyHandler("EnableTargetLocking", true));
+                AddBooleanHandler(Property.USE, b => b.IsUnderControl);
+                AddBooleanHandler(Property.ROLL_INPUT, b => b.EnableIdleRotation, (b,v) => b.EnableIdleRotation = v);
+                AddReturnHandlers(Property.TARGET, Return.VECTOR,
+                    TypeHandler(VectorHandler(GetTarget, SetTarget), Return.VECTOR),
+                    TypeHandler(BooleanHandler(b => b.HasTarget, (b, v) => { if (!v) ResetTarget(b); }), Return.BOOLEAN),
+                    TypeHandler(StringHandler(b => b.GetTargetingGroup(), (b, v) => b.SetTargetingGroup(v)), Return.STRING));
+                AddVectorHandler(Property.TARGET_VELOCITY, b => b.GetTargetedEntity().Velocity, (b, v) => b.TrackTarget(GetTarget(b), v));
+                AddNumericHandler(Property.ANGLE, b => b.Azimuth * RadiansToDegrees, (b, v) => b.Azimuth = v * DegreesToRadians, 5);
+                AddNumericHandler(Property.ALTITUDE, b => b.Elevation * RadiansToDegrees, (b, v) => b.Elevation = v * DegreesToRadians, 5);
```
Comment (jgersti) [inline EasyCommands/BlockHandlers/TurretBlockHandlers.cs:65]: azimuth and elevation needs to synchronized (`SyncAzimuth()`, `SyncElevation()`) if i read the mdk wiki right.

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                return tools.Where(t => t is IMyUserControllableGun).Select(t => (IMyUserControllableGun)t);
+            }
+        }
+
+        public class TurretBlockHandler<T> : GunBlockHandler<IMyLargeTurretBase> {
+            public TurretBlockHandler() {
+                AddNumericHandler(Property.RANGE, b => b.Range, (b,v) => b.Range = v, 100f);
+                AddPropertyHandler(Property.LOCKED, TerminalPropertyHandler("EnableTargetLocking", true));
+                AddBooleanHandler(Property.USE, b => b.IsUnderControl);
+                AddBooleanHandler(Property.ROLL_INPUT, b => b.EnableIdleRotation, (b,v) => b.EnableIdleRotation = v);
```
Comment (jgersti) [inline EasyCommands/BlockHandlers/TurretBlockHandlers.cs:58]: idle movement needs to be synchronized (`SyncEnableIdleRotation`).

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+            Vector3D GetTarget(IMyLargeTurretBase turret) =>
+                GetVector(GetCustomProperty(turret, "target") ?? "") ?? (turret.HasTarget ? GetPosition(turret.GetTargetedEntity()) : Vector3D.Zero);
+
+            void ResetTarget(IMyLargeTurretBase turret) {
+                //Idle Movement setting gets reset when calling ResetTargetingToDefault(), so need to re-apply it.
+                bool idleMovement = turret.EnableIdleRotation;
+                turret.ResetTargetingToDefault();
+                DeleteCustomProperty(turret, "target");
+                turret.EnableIdleRotation = idleMovement;
```
Comment (jgersti) [inline EasyCommands/BlockHandlers/TurretBlockHandlers.cs:78]: see above

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            public TurretControlBlockHandler() {
+                AddNumericHandler(Property.RANGE, b => b.Range, (b, v) => b.Range = v, 100f);
+                AddPropertyHandler(Property.LOCKED, TerminalPropertyHandler("EnableTargetLocking", true));
+                AddPropertyHandler(Property.AUTO, TerminalPropertyHandler("AI", true));
+                AddBooleanHandler(Property.USE, b => b.IsUnderControl);
+                AddReturnHandlers(Property.TARGET, Return.VECTOR,
+                    TypeHandler(VectorHandler(b => {
+                        Vector3D? target = Vector3D.Zero;
+                        if (b.HasTarget) target = GetPosition(b.GetTargetedEntity());
+                        return target.Value;
```
Comment (jgersti) [inline EasyCommands/BlockHandlers/TurretBlockHandlers.cs:32]: I am a bit confused why that is not simply 
```
b => b.HasTarget ? GetPosition(b.GetTargetEntity()) : Vector3D.Zero;
```

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs (hunk 1 above)
Comment (MerlinofMines) [reply EasyCommands/BlockHandlers/TurretBlockHandlers.cs:49]: good call :)

File: EasyCommands/BlockHandlers/TurretBlockHandlers.cs (hunk 5 above)
Comment (MerlinofMines) [reply EasyCommands/BlockHandlers/TurretBlockHandlers.cs:32]: I copied from other and didn't consider that it can be simplified, good callout.

**Pull request merged**